completion_aggregator/management/commands/migrate_progress.py

```diff
@@ -0,0 +1,165 @@
+            batch_size = min(options['batch_size'], stop - index)
+            migrate_batch.apply_async(
+                kwargs={'offset': index, 'batch_size': batch_size},
+                **task_options
+            )
+
+    def get_task_options(self, options):
+        """
+        Return task options for generated celery tasks.
+
+        Currently, this adds a routing key, if provided.
+        """
+        opts = {}
+        if options.get('routing_key'):
+            opts['routing_key'] = options['routing_key']
+        return opts
+
+    def _set_log_level(self, options):
+        """
+        Sets logging levels for this module and the block structure
+        cache module, based on the given the options.
+        """
+        if options.get('verbosity') == 0:
+            log_level = logging.WARNING
+        elif options.get('verbosity') >= 1:
+            log_level = logging.INFO
+        log.setLevel(log_level)
+
+
+@celery.shared_task
+def migrate_batch(offset, batch_size):  # Cannot pass a queryset to a task.
+    """
+    Convert a batch of CourseModuleCompletions to BlockCompletions.
+
+    Given an offset and batch_size, this task will:
+
+    * Fetch a subset of the existing CourseModuleCompletions,
+    * Update the BlockCompletions table
+    * Run aggregators for each one.
+    """
+
+    # Keep the ordering consistent with the index.
+    queryset = CourseModuleCompletion.objects.all().order_by('course_id', 'content_id', 'user')
+    course_module_completions = queryset[offset:offset + batch_size]
+
+    processed = {}  # Dict has format: {course: {user: [blocks]}
+    insert_params = []
+    for cmc in course_module_completions:
+        try:
+            course_key = CourseKey.from_string(cmc.course_id)
+            block_key = UsageKey.from_string(cmc.content_id).map_into_course(course_key)
+            block_type = block_key.block_type
+        except InvalidKeyError:
+            log.exception(
+                "Could not migrate CourseModuleCompletion with values: %s",
+                cmc.__dict__,
+            )
+            continue
+        if course_key not in processed:
+            processed[course_key] = {}
+        if cmc.user not in processed[course_key]:
+            processed[course_key][cmc.user] = set()
+        processed[course_key][cmc.user].add(block_key)
+        # Param order: (user_id, course_key, block_key, block_type)
+        insert_params.append((cmc.user_id, cmc.course_id, cmc.content_id, block_type))
+    if connection.vendor == 'mysql':
+        sql = INSERT_OR_UPDATE_MYSQL
+    else:
+        sql = INSERT_OR_UPDATE_SQLITE
+    with connection.cursor() as cur:
+        cur.executemany(sql, insert_params)
+    for course_key in processed:
+        for user in processed[course_key]:
+            blocks = processed[course_key][user]
+            # Run synchronously.
+            tasks._update_aggregators(user=user, course_key=course_key, block_keys=blocks)  # pylint: disable=protected-access
```

Comment:
Do we want to retry if there are failures or do you want to target a specific batch instead?
What happens if we update the table but fail to run _update_aggregators for some reason (task failure in the loop, etc).  Do we handle that retry by just rerunning the whole batch because it's safe?
Should we think about filing the tasks from the management command with a delay between them?  I don't have a good sense of scale of the tables in play here.

The SQL looks fine, aside from me wondering if we want a transaction here in some way, especially if some already processed learners/blocks will start writing to the new table while we're migrating?

Reply:
Doesn't matter.  I added a `--delay-between-tasks` argument, so you can tweak that as needed.

Reply:
I'll re-read tomorrow (sorry, on-call dominating my week).
Transaction sounds good.
I think the delay-between-tasks works well.  Will you also make which celery queue we target configurable?  That way we can silo the work off to the side and not accidentally block any of the other things they do (or limit workers on that queue for similar reasons).  Especially relevant if this gets used on larger installs.

I can go poking in the database to look at sizes, let me know which other tables may make sense because the coursemodulecompletions one

Reply:
I've already got a --routing-key argument.  See LL59-63, and LL104-13 and LL93-101 IIRC that's how we did it when we did the Big Grades Backfill of 2017.

Reply:
I think as far as this script is concerned, performance will be primarily affected by the size of the progress_coursemodulecompletion table.

Reply:
Ah, yes, I saw routing key the first time I read this, but last night I was looking for the decorator.  That works.
I'm going to examine that table size in a few places and will reread but then I think this should be good from our perspective